proof/refine/ARM/Ipc_R.thy

```diff
@@ -2522,18 +2522,499 @@ global_interpretation setMessageInfo: typ_at_all_props' "setMessageInfo t info"
+                  apply (clarsimp simp: tcb_cte_cases_def)
+                 apply (clarsimp simp: sc_relation_def)
+                apply (wpsimp wp: hoare_drop_imps)
+               apply (wpsimp wp: hoare_drop_imps
+                                 threadSet_valid_release_queue threadSet_valid_release_queue'
+                                 threadSet_valid_queues_no_state threadSet_valid_queues'_no_state)
+              apply (rule_tac x="the (sc_tcb sc)" and x'="the (scTCB sca)" in lift_args_corres)
+               apply (rule tcb_release_remove_corres)
+              apply (clarsimp simp: sc_relation_def)
+             apply (wpsimp | strengthen weak_valid_sched_action_strg)+
+            apply (rule_tac Q="\<lambda>_. tcb_at' (the (scTCB sca)) and valid_tcbs' and
+                                   valid_queues and valid_queues' and
+                                   valid_release_queue and valid_release_queue' and
+                                   (\<lambda>s. \<forall>d p. the (scTCB sca) \<notin> set (ksReadyQueues s (d, p)))"
+                         in hoare_strengthen_post[rotated])
+             apply (clarsimp simp: valid_release_queue'_def obj_at'_def)
+            apply (wpsimp wp: tcbReleaseRemove_valid_queues hoare_vcg_all_lift)
+           apply (rule_tac x="the (sc_tcb sc)" and x'="the (scTCB sca)" in lift_args_corres)
+            apply (rule tcbSchedDequeue_corres)
+           apply (clarsimp simp: sc_relation_def)
+          apply (wpsimp wp: hoare_vcg_all_lift hoare_vcg_imp_lift')
+         apply (wpsimp wp: tcbSchedDequeue_valid_queues hoare_vcg_all_lift)
+        apply wpsimp+
+    apply (frule invs_valid_objs)
+    apply (fastforce simp: valid_obj_def valid_sched_context_def valid_bound_obj_def obj_at_def)
+   apply (prop_tac "sc_at' scp s' \<and> tcb_at' thread s'")
+    apply (fastforce elim: sc_at_cross tcb_at_cross simp: state_relation_def)
+   apply clarsimp
+   apply (frule valid_objs'_valid_tcbs')
+   apply (rule valid_objsE', assumption)
+    apply (fastforce simp: obj_at'_def projectKO_eq projectKO_sc)
+   apply (clarsimp simp: valid_obj'_def valid_sched_context'_def obj_at'_def projectKOs)
+   apply (frule valid_objs'_valid_tcbs')
+   apply (fastforce simp: valid_obj'_def valid_tcb'_def)
+  apply (clarsimp simp: sym_refs_asrt_def)
+  done
+
+lemma maybeDonateSc_corres:
+  "corres dc (tcb_at tcb_ptr and ntfn_at ntfn_ptr and invs and weak_valid_sched_action)
+             (tcb_at' tcb_ptr and ntfn_at' ntfn_ptr and (\<lambda>s. valid_objs' s \<and>
+                              valid_queues s \<and>
+                              valid_queues' s \<and>
+                              valid_release_queue s \<and>
+                              valid_release_queue' s))
+            (maybe_donate_sc tcb_ptr ntfn_ptr)
+            (maybeDonateSc tcb_ptr ntfn_ptr)"
+  unfolding maybeDonateSc_def maybe_donate_sc_def
+  apply (simp add: get_tcb_obj_ref_def get_ntfn_obj_ref_def2 liftM_def maybeM_def
```

Comment:
I haven’t loaded up in Isabelle, so I might be wrong, but it looks like simplifying with `get_ntfn_obj_ref_def2 liftM_def` is the same thing as simply unfolding `get_ntfn_obj_ref_def`? Or am I missing something? Why do we need `get_ntfn_obj_ref_def2` and `get_sc_obj_ref_def2`?

Reply:
You're right. Both of these can go.